directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,268 @@
+package directorywatcher
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	"gopkg.in/fsnotify.v1"
+
+	"github.com/reddit/baseplate.go/filewatcher"
+	"github.com/reddit/baseplate.go/internal/limitopen"
+	"github.com/reddit/baseplate.go/log"
+)
+
+// AddFile is a type of function that should be ran to handle
+// adding data after a file has been parsed
+type AddFile func(d interface{}, file interface{}) (data interface{}, err error)
```

Comment:
I have a hard time wrapping my head around this function without reading further down how it is being utilized, part of this is the constraint of the empty interface being unconcrete here but I also wonder if we can improve it with either different parameter names or more document.

Is `d` the existing data so far and file is the content of a parsed file that will be merged into the existing data `?` Maybe we can clarify this.